### ISSUE TITLE
How to extract Feature from a Image Loaded by OpenCV

### ISSUE THREAD

**Issue by RisithPerera** [layumi/Person_reID_baseline_pytorch#256]
Hey, I have successfully extracted the features from an image that is loaded by cv2.imread(). Then I try to load that same Image using Pytorch ImageFolder and Dataloader and got the features. But the feature values are not the same for both.  So my score value is around 0.2. It should be 1.0 for the same person image.  I think because both image pixel values are not the same after applying the transform. We need to convert the OpenCV image into a PIL image. Actually, I want to use this in a video. So then I can track people. Here I have attached my test code.

`
        from __future__ import print_function, division
        import torch
        import torch.nn as nn
        from torch.autograd import Variable
        from torchvision import datasets, transforms
        from model import ft_net
        import cv2

    model_path = "./model/ft_ResNet50/net_last.pth"
    stride = 2
    batchsize = 256
    nclasses = 751
    gpu_id = 0

    def image_feature(model, img):
        feature = torch.FloatTensor()
        n, c, h, w = img.size()
        ff = torch.FloatTensor(n, 512).zero_().cuda()
        for i in range(2):
            if (i == 1):  # Flip the image
                inv_idx = torch.arange(img.size(3) - 1, -1, -1).long()
                img = img.index_select(3, inv_idx)

            input_img = Variable(img.cuda())
            outputs = model(input_img)
            ff += outputs

        fnorm = torch.norm(ff, p=2, dim=1, keepdim=True)
        ff = ff.div(fnorm.expand_as(ff))
        feature = torch.cat((feature, ff.data.cpu()), 0)
        return feature

    def loader_feature(model, dataloader):
        features = torch.FloatTensor()
        for data in dataloader:
            img, label = data
            n, c, h, w = img.size()
            ff = torch.FloatTensor(n,512).zero_().cuda()
            for i in range(2):
                if(i==1): #Flip the image
                    inv_idx = torch.arange(img.size(3) - 1, -1, -1).long()
                    img = img.index_select(3, inv_idx)

                input_img = Variable(img.cuda())
                outputs = model(input_img)
                ff += outputs

            fnorm = torch.norm(ff, p=2, dim=1, keepdim=True)
            ff = ff.div(fnorm.expand_as(ff))
            features = torch.cat((features,ff.data.cpu()), 0)
        return features

    if __name__ == '__main__':
        #for Images loaded by OpenCV
        transImage = transforms.Compose([
            transforms.ToPILImage(),  # Added by my self
            transforms.Resize((256, 128), interpolation=3),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        ])

        # for Images loaded by DataLoader
        transLoader = transforms.Compose([
            transforms.Resize((256, 128), interpolation=3),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        ])

        image = cv2.imread('./database/0001/0001_001.png')
        image = transImage(image).unsqueeze(0)

        image_dataset = datasets.ImageFolder('./database/', transLoader)
        image_loader = torch.utils.data.DataLoader(image_dataset, batch_size=batchsize, shuffle=False, num_workers=0)

        model = ft_net(nclasses, stride=stride)
        model.load_state_dict(torch.load(model_path))
        model.classifier.classifier = nn.Sequential()

        # Change to test mode
        model = model.eval()
        if torch.cuda.is_available():
            model = model.cuda()

        with torch.no_grad():
            load_feature = loader_feature(model,image_loader) # Images loaded by Dataloader
            img_feature = image_feature(model, image)         # Images read by OpenCV

        print(f'DataLoader Feature Shape: {load_feature.numpy().shape}')
        print(load_feature.numpy()[0][:10])  # Print only first 10 values out of 512

        print(f'OpenCV Image Feature Shape: {img_feature.numpy().shape}')
        print(img_feature.numpy()[0][:10])  # Print only first 10 values out of 512

        query = img_feature.view(-1, 1)
        score = torch.mm(load_feature, query)
        score = score.squeeze(1).cpu()
        print(f'Score: {score.numpy()}')`

**Comment by layumi** [comment]
Hi @RisithPerera 

Opencv provide BGR? which is different from RGB. (I used opencv a few years ago. I am not sure whether channels are aligned.  

You may refer to : 
https://stackoverflow.com/questions/43232813/convert-opencv-image-format-to-pil-image-format
